### PR TITLE
Pull request for issue #20

### PR DESCRIPTION
**Pull request by inthar-raven**

### CODE REVIEW

**Comment by frostburn** [comment]
Here's what I get from the benchmark:
```
Bjorklund's algorithm x 23,327 ops/sec ±1.91% (90 runs sampled)
Bresenham line algorithm x 128,019 ops/sec ±3.26% (85 runs sampled)
Optimized Bjorklund's algorithm x 46,776 ops/sec ±1.01% (92 runs sampled)
Fastest is Bresenham line algorithm
```

File: src/helpers.ts [hunk 1]
```diff
@@ -1,9 +1,83 @@
...
+ * Distribute subsequences as evenly as possible using Björklund's algorithm;
+ * modified as to always return the brightest mode.
+ */
+export function bjorklund_brightest_bool(a: number, b: number): boolean[] {
+  if (isNaN(a) || isNaN(b)) {
+    throw new Error('Invalid input');
+  }
+  const d = gcd(a, b);
+  if (d === 1) {
+    let [countFirst, countSecond] = [a / d, b / d];
```
Comment (frostburn) [inline src/helpers.ts:35]: If `d` equals `1` you don't need to divide by it.

File: src/helpers.ts [hunk 2]
```diff
@@ -1,9 +1,83 @@
...
+      else {
+        countSecond = countSecond - countFirst;
+        first = first.concat(second);
+      }
+      // At the current step we have `countFirst` `first` substrings and `countSecond` `second` substrings,
+      // where we must guarantee that `first` is lexicographically greater than `second`.
+      // Thus if first > second, then switch them and the count variables.
+      // Do this step before checking the while condition; we know the desired lex. ordering holds for the first step,
+      // and our stopping condition requires that `first` < `second` actually hold to really behave correctly.
+      if (!lexCompareArrays<boolean>(first, second)) {
```
Comment (frostburn) [inline src/helpers.ts:58]: TypeScript is fairly smart with inference. I don't think explicit `<boolean>` is required here.

File: src/helpers.ts [hunk 3]
```diff
@@ -1,9 +1,83 @@
...
+    return Array(countFirst).fill(first).flat().concat(second);
+  } else {
+    return Array(d)
+      .fill(bjorklund_brightest_bool(a / d, b / d))
+      .flat();
+  }
+}
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+// @ts-ignore
+export function bjorklund_brightest<T>(
```
Comment (frostburn) [inline src/helpers.ts:72]: I don't think we need this generic version. It's used in a boolean context anyway.

File: src/helpers.ts [hunk 4]
```diff
@@ -1,9 +1,83 @@
...
+    }
+    return Array(countFirst).fill(first).flat().concat(second);
+  } else {
+    return Array(d)
+      .fill(bjorklund_brightest_bool(a / d, b / d))
+      .flat();
+  }
+}
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+// @ts-ignore
```
Comment (frostburn) [inline src/helpers.ts:71]: What is this? Only sub-par third-party libraries get a pass. Our code should adhere to the type and style guide.

File: src/helpers.ts [hunk 5]
```diff
@@ -1,9 +1,83 @@
...
+export function bjorklund_brightest<T>(
+  a: number,
+  b: number,
+  first: T,
+  second: T
+): T[] {
+  return bjorklund_brightest_bool(a, b).map(x => (!x ? first : second));
```
Comment (frostburn) [inline src/helpers.ts:78]: If you decide to keep it. `x ? second : first` saves a character.

**Comment by frostburn** [comment]
What's up with `package-lock.json`?

**Comment by inthar-raven** [comment]
> Here's what I get from the benchmark:
> 
> ```
> Bjorklund's algorithm x 23,327 ops/sec ±1.91% (90 runs sampled)
> Bresenham line algorithm x 128,019 ops/sec ±3.26% (85 runs sampled)
> Optimized Bjorklund's algorithm x 46,776 ops/sec ±1.01% (92 runs sampled)
> Fastest is Bresenham line algorithm
> ```

I wonder why my version of Bjorklund is still 3 times as slow as Bresenham. It might be due to all the MOS subword comparisons I used to get the brightest mode are slowing Bjorklund down, whereas Bresenham just does one pass through the brightest word.

**Comment by inthar-raven** [comment]
I just removved the old bjorklund() function.

**Comment by frostburn** [comment]
The computed mode doesn't have to be the brightest. Just not the darkest so that the generator can be located at the expected degree.

**Comment by frostburn** [comment]
Current benchmark results:
```
Bjorklund's algorithm x 46,449 ops/sec ±0.88% (90 runs sampled)
Bresenham line algorithm x 134,099 ops/sec ±1.26% (85 runs sampled)
Fastest is Bresenham line algorithm
```

**Comment by inthar-raven** [comment]
I forgot to mention there's another array operation my Bjorklund relies on:
```ts
        [first, second] = [first.concat(second), first];
```
And array operations are just slow.

**Comment by frostburn** [comment]
One idea could be to go full lexicographic and try Björklund's with strings.

**Comment by inthar-raven** [comment]
Oh, wow!
```
Bjorklund's algorithm x 414,195 ops/sec ±0.66% (90 runs sampled)
Bresenham line algorithm x 165,760 ops/sec ±1.18% (91 runs sampled)
Fastest is Bjorklund's algorithm
```

**Comment by frostburn** [comment]
Could you not include `package-lock.json`? I'm failing to reproduce those changes on my end.

**Comment by frostburn** [comment]
Once you're done tweaking, remember to squash your commits so that irrelevant churn isn't included in the history on the `main` branch.

**Comment by inthar-raven** [comment]
> Could you not include `package-lock.json`? I'm failing to reproduce those changes on my end.

I removed package-lock.json but it seems to be required.

**Comment by frostburn** [comment]
> > Could you not include `package-lock.json`? I'm failing to reproduce those changes on my end.
> 
> I removed package-lock.json but it seems to be required.

Sorry, I meant not include your changes to it. Just use the one on `origin/main`.

**Pull request merged**